### ISSUE TITLE
Issues / vulnerability with E2E encryption

### ISSUE THREAD

**Issue by ieure** [alphapapa/matrix-client.el#90]
Summary:

 - Messages sent to E2E encrypted rooms are sent cleartext with no warning or indication of what’s happening.
 - Encrypted messages sent to E2E encrypted rooms are silently discarded.

After installing matrix-client.el, I found that some rooms weren't working, while others were.  In the broken rooms, I could send messages and see typing notifications, but new messages never appeared.  The missing messages show up in Riot, so this is a matrix-client.el issue.

Eventually, I realized that the difference between the working and non-working rooms was whether E2E encryption was enabled.  On rooms with E2E encryption, matrix-client.el silently discards any message except my own.

This got me more curious, so I looked at the message detail in Riot, comparing messages I sent with Riot to ones I sent with matrix-client.el.  While the Riot messages are encrypted (I can click the "..." at right and see "End-to-end encryption information" in the dialog that pops up), messages sent from matrix-client.el aren’t encrypted at all, even though the room is set to E2E encryption.  This deeply violates my expectations about Matrix client behavior -- if a room is encrypted, cleartext messages should **NEVER** be sent to it.  I’m actually surprised that the Matrix server allowed it to happen.

**Comment by alphapapa** [comment]
Yes, this client doesn't support encryption, and it probably never will, given the complexities involved and the difficulty integrating it into Emacs with a C module.  And since I don't use E2E rooms, I'm not very interested in working on it.

I agree that the server ought to reject unencrypted messages to encrypted rooms, but I'd be surprised if the Matrix folks hadn't already considered that, so maybe there's a reason that's the case.  Have you looked at the spec and discussions about it?  I'm sure it's been discussed before.

Yes, unrecognized messages aren't displayed.  Matrix is a relatively free-form protocol, and clients can send whatever "types" of events they want to a room.  It wouldn't be too difficult to display placeholders for unrecognized event types, so I'm open to a PR for that.

I'm also open to a PR to disable sending messages to encrypted rooms, but I don't know how that part of the spec works.

AFAIK many Matrix clients don't support E2E, and I don't know how they handle E2E rooms.

It does seem like a problem in the spec to me.  E2E isn't integral to rooms, so clients can simply ignore it, and the server doesn't bother to enforce it.  But that may be why servers don't enforce it: there's no way for the network as a whole to enforce the enforcement on servers.  A server could ignore such a requirement, and so can any client.

**Comment by alphapapa** [comment]
To be clear, I would like to avoid sending unencrypted messages to rooms the user expects to be encrypted.  I don't want to do the wrong or unexpected thing.  So I'm looking at the spec to determine how to determine if a room is supposed to be encrypted.

The 0.3.0 spec, which is basically what this client targets, doesn't explain how.

The 0.4.0 spec doesn't seem to explain either.  It seems to say that a client is expected to determine whether other users have public keys that should be used when sending messages.  For rooms, from what I can understand, I infer that a room's `state` events must be inspected to see if one of them is of the type `m.room.encryption`.

That seems straightforward enough, although I wish there were a boolean flag maintained by the server.  Inspecting every room event seems awkward, since some rooms have many (e.g. an abbreviated list of the number of state events for rooms I'm currently in: `(1800 630 24 1383 20 558 9 7 39 1835 7 11 ...)`).

So I guess that means that, when processing all of those events as they arrive (and on initial sync), we need to set our own client-side flag if we encounter an event of that type.  Then, if the user tries to send a message to a room with that flag, we signal an error instead of sending the message.  We can also insert a message into the room's buffer warning that it's an encrypted room, which is unsupported.

What do you think?

**Comment by alphapapa** [comment]
@ieure Please try this branch: https://github.com/alphapapa/matrix-client.el/tree/wip/warn-about-encrypted-rooms  It should display a warning in the header of encrypted rooms and refuse to send messages to them.

**Comment by alphapapa** [comment]
@ieure Ping?  :)  If you can verify that that branch works as intended (i.e. doesn't permit sending unencrypted messages to encrypted rooms, and displays the warning in the header for encrypted rooms), then I'll merge it.

**Comment by seanfarley** [comment]
> given the complexities involved and the difficulty integrating it into Emacs with a C module

I just started to use matrix and found this issue :-) I have some experience with integrating C modules, so I'm curious what the complexities are. To be clear, I have zero experience using the `libolm` so I'm assuming it works in some simple way (e.g. a function to encrypt a given message, and a function to decrypt an incoming message)?

**Comment by non-Jedi** [comment]
Not even close. See, for example, [the end-to-end encryption implementation guide](https://matrix.org/docs/guides/end-to-end-encryption-implementation-guide/).

**Comment by seanfarley** [comment]
Holy ...

![E85E1B21-04D2-477E-9F38-5853342021F5](https://user-images.githubusercontent.com/20887/84558548-87441e80-ace8-11ea-894b-d3a0d814e1a3.jpeg)

**Comment by alphapapa** [comment]
@seanfarley LOL  :)

If someone who uses E2EE with Matrix would be willing to test that branch, it would be helpful, so at least we'd display appropriate warnings.

**Comment by seanfarley** [comment]
@alphapapa I needed to rebase (to fix the `void-variable` error) and fix up a few typos: https://github.com/alphapapa/matrix-client.el/compare/wip/warn-about-encrypted-rooms...seanfarley:wip/warn-about-encrypted-rooms

Not sure what the `nil: nil` is supposed to mean but the warnings stopped me from sending text (cool!).

Screenshots of the UI:
<img width="498" alt="Screen Shot 2020-06-14 at 12 49 01 PM" src="https://user-images.githubusercontent.com/20887/84602720-3e05e300-ae3e-11ea-8b36-6c4ce71351bd.png">

<img width="456" alt="Screen Shot 2020-06-14 at 12 48 52 PM" src="https://user-images.githubusercontent.com/20887/84602713-334b4e00-ae3e-11ea-96c4-2c804a0be19a.png">

**Comment by snan** [comment]
https://github.com/skeeto/elisp-ffi ← this any good? or one could use it's implementation strategy but with one of the really FFI-friendly schemes or lisps like Chicken.

**Comment by alphapapa** [comment]
@snan Something like that might be possible, but I'm not interested in adding that kind of complexity to this project.

**Comment by alphapapa** [comment]
Closing issues and archiving repository.  Please see https://github.com/alphapapa/ement.el for the new client.
